File: backend/src/algo/dfs.cpp
#include "algo.hpp"

#include <functional>
#include <set>
#include <utility>

namespace tour::algo {

// DFS 生成导游线路图，语义见 API.md 3.1：
// 邻点按 id 升序访问；walk 记录含回溯的完整行走序列；
// edges 是 walk 相邻两点组成的有向边（去重）；没走到的景点放 unreachable
TourResult buildTour(const Graph& g, int startId) {
    if (!g.hasVertex(startId)) throw std::invalid_argument("buildTour: 起点不存在");

    TourResult result;
    result.startId = startId;

    std::set<int> visited;
    std::vector<int>& walk = result.walk;

    // 用 std::function 是因为递归 lambda 要能调用到它自己
    std::function<void(int)> dfs = [&](int u) {
        visited.insert(u);
        walk.push_back(u);
        for (const auto& nb : g.neighbors(u)) {  // neighbors 已按邻点 id 升序排好
            int v = nb.first;
            if (visited.count(v)) continue;
            dfs(v);
            walk.push_back(u);  // 从 v 退回到 u，也算一步
        }
    };
    dfs(startId);

    // 相邻两点组成有向边，按第一次出现的顺序去重
    std::set<std::pair<int, int>> seen;
    for (size_t i = 0; i + 1 < walk.size(); i++) {
        if (seen.insert({walk[i], walk[i + 1]}).second) {
            result.edges.push_back({walk[i], walk[i + 1]});
        }
    }

    // 没被访问到的就是 unreachable（vertexIds 本身升序）
    for (int id : g.vertexIds()) {
        if (!visited.count(id)) result.unreachable.push_back(id);
    }
    return result;
}

}  // namespace tour::algo

File: frontend/src/algo/dfs.js
import { buildAdjacency, vertexIds } from './graph'

/**
 * 深度优先策略生成导游线路图。
 * 语义（与 docs/API.md 3.1 一致）：
 *  - 邻点按景点 id 升序访问
 *  - walk 为含回溯的完整行走序列
 *  - edges 为 walk 相邻点构成的有向边（含回溯反向边），去重后输出，即导游线路图
 * 返回 { startId, walk, edges, unreachable }
 */
export function buildTour(attractions, edges, startId) {
  const ids = vertexIds(attractions)
  if (!ids.includes(startId)) {
    throw new Error(`景点不存在: ${startId}`)
  }
  const adj = buildAdjacency(attractions, edges)
  const visited = new Set()
  const walk = []

  const dfs = (u) => {
    visited.add(u)
    walk.push(u)
    for (const { to: v } of adj.get(u)) {
      if (visited.has(v)) continue
      dfs(v)
      walk.push(u)
    }
  }
  dfs(startId)

  const seenEdge = new Set()
  const tourEdges = []
  for (let i = 0; i + 1 < walk.length; i++) {
    const key = `${walk[i]}->${walk[i + 1]}`
    if (!seenEdge.has(key)) {
      seenEdge.add(key)
      tourEdges.push({ from: walk[i], to: walk[i + 1] })
    }
  }

  return {
    startId,
    walk,
    edges: tourEdges,
    unreachable: ids.filter((id) => !visited.has(id)),
  }
}
